crates/ibc/src/core/entrypoint.rs

```diff
@@ -0,0 +1,42 @@
+#[cfg(feature = "val-exec-entry")]
+use ibc_proto::google::protobuf::Any;
+
+#[cfg(feature = "val-exec-entry")]
+use super::ValidationContext;
+
+use super::{ExecutionContext, MsgEnvelope, RouterError};
+
+use crate::core::handler::{execution_handler, validation_handler};
+
+/// Entrypoint which performs both validation and message execution
+pub fn dispatch(ctx: &mut impl ExecutionContext, msg: MsgEnvelope) -> Result<(), RouterError> {
+    validation_handler(ctx, msg.clone())?;
+    execution_handler(ctx, msg)
+}
+
+/// Entrypoint which only performs message validation
+///
+/// If a transaction contains `n` messages `m_1` ... `m_n`, then
+/// they MUST be processed as follows:
+///     validate(m_1), execute(m_1), ..., validate(m_n), execute(m_n)
+/// That is, the state transition of message `i` must be applied before
+/// message `i+1` is validated. This is equivalent to calling
+/// `dispatch()` on each successively.
+#[cfg(feature = "val-exec-entry")]
+pub fn validate<Ctx>(ctx: &Ctx, message: Any) -> Result<(), RouterError>
```

Comment:
I think we should have all entry points use `MsgEnvelope`. I don't like `Any` because it hides the real type, is arbitrary (why `Any` as opposed to simply using the actual `MsgEnvelope` that we expect?), and if users actually have a `MsgEnvelope` in an `Any`, they only need to add 1 line of code before calling our entrypoints.

Reply:
[4d1ca0a](https://github.com/cosmos/ibc-rs/pull/661/commits/4d1ca0a825b2f37be264c55e2cc9caf5bead0b20)